### PR TITLE
Provision documentation

### PR DESCRIPTION
**Pull request by iomodo**
#### Summary
This PR adds provisioning documentation

### CODE REVIEW

**Comment by hanzei** [comment]
@iomodo The diff looks weird. Should this branch target a different base then `master`?

**Comment by hanzei** [comment]
The real changes seem to live in [`c5e2081` (#104)](https://github.com/mattermost/mattermost-plugin-apps/pull/104/commits/c5e2081bc6e9757713e309988a23a279d6f3be71), right?

**Comment by levb** [comment]
@hanzei @iomodo fixed: the conflicts appeared when I merged `lev-docs` into `master`, so I CP-ed the commit and force-pushed.

File: docs/07-lifecycle.md
Comment (levb) [inline docs/07-lifecycle.md:6]: We should probably have an AWS version of Hello World manifest here

File: docs/07-lifecycle.md
Comment (levb) [inline docs/07-lifecycle.md:42]: we need to simplify terminology - published, registered, listed, (or else?)

**Review by levb — CHANGES_REQUESTED** [review]

File: docs/07-lifecycle.md
Comment (levb) [reply docs/07-lifecycle.md:42]: @iomodo ^^ is mostly for @justinegeffen to edit away, optional for this PR AFAIC.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:6]: We should prob. start with stating the purpose of the app bundle, what process uses it, etc. before describing what it is composed of.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:7]: we should prob. call out if these files are ALL the static files an app can have/needs (or does it depend if it is an http vs. AWS hosted app, etc. ).

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:8]: worth being specific on the "information".

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:9]: Is there a versioned specification/structure for lambda functions? (like the ones in the bundle should respect format version X).

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:12]: we should prob. detail which appsctl command does the actual provisioning  and call out appsctl as a tool/script, etc.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:24]: "semantic names" - example? and example of mapping to AWS name.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:32]: can we detail more about the "semantic keys" in this context?

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:38]: "(the same way plugins do right now)" - propose we remove this - the audience will be the developer that might not have our context.
"On a new app release bundle" - propose to describe in more depth the process to get here.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:42]: apps.json -> manifests.json?

Comment (catalintomai) [inline docs/07-lifecycle.md:42]: "On a totally new app registration or on a registration of the new version of the already registered app, a new version of the Proxy Plugin is cut" - is this indeed the case?

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:53]: I believe we should also add some diagrams to describe the various flows.

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:65]: move (and expand) the upgrade discussion here? calling out not tackling downgrade?

Comment (levb) [reply docs/07-lifecycle.md:65]: I think this should be a separate ticket in Jira. The `master` implementation now sends `OnVersionChanged` message, and I propose that we only include the Bot credentials in it; Bot is sufficient to migrate the KV values stored; and to send users messages if they need to do something requiring user tokens.

I am 1/5 that we should scrape the "handling of upgrades/downgrades" altogether until we have a more solid use-case.

File: docs/07-lifecycle.md
Comment (iomodo) [reply docs/07-lifecycle.md:6]: We already have the hello manifest in the docs right? Should I duplicate?

File: docs/07-lifecycle.md
Comment (iomodo) [reply docs/07-lifecycle.md:42]: We have to change the `manifests.json` file so it will be tied to the PP release

File: docs/07-lifecycle.md
Comment (iomodo) [reply docs/07-lifecycle.md:38]: I think app release process could be determined elsewhere

File: docs/07-lifecycle.md
Comment (catalintomai) [inline docs/07-lifecycle.md:62]: Should we also mention the "builtin" apps here (usage, flows, etc.)?

**Review by catalintomai — APPROVED** [review]
cc: @justinegeffen

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:10]:
```suggestion
- Each of the lambda function bundles is a valid and runnable AWS Lambda function, provisioned in AWS by the [Mattermost Apps Cloud Deployer](https://github.com/mattermost/mattermost-apps-cloud-deployer).
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:7]:
```suggestion
The app bundle contains a `manifest.json` file, a `static/` folder, and one or several lambda function bundles.
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:8]:
```suggestion
- The `static/` folder contains all the static files the app needs. For the Mattermost AWS apps static files are automatically provisioned and stored in the dedicated AWS S3 bucket (`mattermost-apps-bucket-production`). 

Apps have unlimited access to them by providing the static file name to the proxy plugin. For the third-party hosted AWS apps, static files are stored in a different S3 bucket (specified by the third-party). For the HTTP apps, when creating a server, the developer should store the static files in the `/static/file_name` relative URL.
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:9]:
```suggestion
- The `manifest.json` file contains details about the app such as appID, appVersion, appType (HTTP or AWS), requested permissions, requested locations, and information about the  functions such as function path, name, runtime, and handler.
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:11]:
```suggestion
The AWS Lambda function bundle is a `.zip` file which contains scripts or compiled programs and their dependencies. Note that it must be less than 50 MB. 

The exact specification of the bundle varies for different runtimes. For example one can see more details for `node.js` bundles [here](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-package.html).
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:13]:
```suggestion
### Provisioning in third-party AWS
```

File: docs/07-lifecycle.md
Comment (justinegeffen) [inline docs/07-lifecycle.md:47]:
```suggestion
To be provisioned in AWS Mattermost Cloud an app bundle is uploaded to the specific S3 bucket (`mattermost-apps-bucket`). On a new app release, a bundle is created by CircleCI and uploaded to S3. 

[Mattermost apps cloud deployer](https://github.com/mattermost/mattermost-apps-cloud-deployer), running as a k8s cron job every hour, detects the S3 upload, creates appropriate lambda functions, assets, and manifest the same way the `appsclt` does for the third-party accounts. 

The deployer needs lambda function names, asset keys, and manifest key to provision the app. It calls the `aws.GetProvisionDataFromFile (/PATH/TO/THE/APP/BUNDLE)` from the proxy plugin to get the provision data. Same data can be generated using the
```

Comment (justinegeffen) [reply docs/07-lifecycle.md:47]: The last sentence is missing a few words.

**Review by justinegeffen — COMMENTED** [review]
Thanks, @iomodo! I'd like to take a second pass at this content once you've taken a look at my review. I suspect I may have missed out some of it as I only edited the changes in the diff. If you're open to me editing the file directly, I am happy to do that.

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:6]: I suggest we go with `app bundle` as `bundle` is quite generic and it's not clear on first sight what it means.

```suggestion
An app bundle is a convenient way to deliver an app to the Mattermost ecosystem. It provides a way to organize code and resources needed for an app to run. An app bundle is created by the developer of the app. Mattermost uses app bundles to provision and install/uninstall apps. 
```

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:7]: Does it have to contain a `static` folder? What is the bare minimum?

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:8]: 0/5 Do we really need to put the bucket name in the docs?

Comment (hanzei) [inline docs/07-lifecycle.md:8]: Do we have a universal way to writing variable in the docs? Something like
```suggestion
- `static/` folder contains all the static files app needs. For the Mattermost AWS Apps static files are automatically provisioned and stored in the dedicated AWS S3 bucket(`mattermost-apps-bucket-production`). Apps can use them whenever they wish by providing the static file name to the Proxy Plugin. For the third-party hosted AWS apps static files are stored in the different S3 bucket(specified by the third-party). For the HTTP Apps, when creating a server, developer should store the static files in the `/static/$FILE_NAME` relative URL.
```

Comment (hanzei) [inline docs/07-lifecycle.md:8]: Are we considering third party AWS apps in this docs? We neither support that nor do we know if we will, right?

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:13]: Should we make this as the "dev way" as this is the main use case right now?

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:16]: Should we recommend `go install` instead? Otherwise we also have to document the `git clone`.

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:53]: Let's clean up the naming here. What does `Publish` mean vs. `Register`? Is that is the same? Or is one about the status of an app release and the other about the process?

Comment (hanzei) [inline docs/07-lifecycle.md:53]: This section seems to have MM staff as the audience. Should we make that clear here and maybe move it do a separate doc?

File: docs/07-lifecycle.md
Comment (hanzei) [reply docs/07-lifecycle.md:79]: I'm with @levb here. Down/upgrade are a hard topic and we don't know good use cases right now.

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:77]: An app can't be uninstalled via the Marketplace.

**Review by hanzei — CHANGES_REQUESTED** [review]

File: docs/07-lifecycle.md
Comment (hanzei) [inline docs/07-lifecycle.md:8]: I'm 3/5 against using the term `Proxy Plugin` as it isn't obvious what the term means. Also, the plugin is more the a proxy. Can we use something more obvious as `Apps Plugin`?

Comment (iomodo) [reply docs/07-lifecycle.md:8]: Correct, we do not support them for now. I'll mention that

File: docs/07-lifecycle.md
Comment (iomodo) [reply docs/07-lifecycle.md:16]: I was under impression that app developer would already have the proxy plugin cloned. Isn't it a case?

File: docs/07-lifecycle.md
Comment (iomodo) [reply docs/07-lifecycle.md:53]: I'll leave it to @justinegeffen

Comment (iomodo) [reply docs/07-lifecycle.md:53]: I'm 0/5

File: docs/07-lifecycle.md
Comment (hanzei) [reply docs/07-lifecycle.md:19]: I don't think we mention that anywhere. Encouraging people to run `go install github.com/mattermost/mattermost-plugin-apps/cmd/appsctl@latest` seems like a good way to get a user starting.

**Review by hanzei — APPROVED** [review]
LGTM :+1:

**Review by levb — APPROVED** [review]

**Comment by levb** [comment]
OT: @hanzei @justinegeffen 1/5 add a GH label for "doc review"? The PRs that affect the (checked-in) docs should update the docs and get a "QA" review from @justinegeffen?

**Comment by hanzei** [comment]
Other repos are using `Editor Review` as the label name. Objections on sticking with the same also for this repo?

**Comment by levb** [comment]
@justinegeffen ping?

**Review by justinegeffen — APPROVED** [review]

**Pull request merged**
